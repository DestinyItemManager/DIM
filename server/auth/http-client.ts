/**
 * Simple HTTP client for Bungie API in Node.js
 * This is a simplified version of DIM's browser HTTP client for use in the MCP server
 */

import type { HttpClientConfig } from 'bungie-api-ts/http';
import { getApiKey } from '../../config/mcp-middleware';
import { authManager } from './bungie-auth';

/**
 * Create a simple HTTP client for bungie-api-ts
 * Compatible with Node.js fetch
 */
export function createBungieHttpClient(apiKey: string, accessToken: string) {
  return async (config: HttpClientConfig) => {
    let url = config.url;
    if (config.params) {
      url = `${url}?${new URLSearchParams(config.params as Record<string, string>).toString()}`;
    }

    const response = await fetch(url, {
      method: config.method || 'GET',
      headers: {
        'X-API-Key': apiKey,
        Authorization: `Bearer ${accessToken}`,
        'Content-Type': 'application/json',
      },
      body: config.body ? JSON.stringify(config.body) : undefined,
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(`HTTP ${response.status}: ${response.statusText} - ${errorText}`);
    }

    return response.json();
  };
}

/**
 * Create an authenticated Bungie API client
 * Handles all auth checks and throws descriptive errors if auth is not set up
 */
export async function createAuthenticatedBungieClient() {
  // Check authentication
  await authManager.initialize();
  const accessToken = await authManager.getAccessToken();
  const membershipId = authManager.getMembershipId();

  if (!accessToken || !membershipId) {
    throw new Error(
      'Not authenticated. A browser window should have opened automatically to DIM. Please log in with your Bungie account. After logging in, tokens will sync automatically. If the browser did not open, manually go to https://localhost:8080 and log in, then try again.',
    );
  }

  // Get API key from browser
  const bungieApiKey = getApiKey();
  if (!bungieApiKey) {
    throw new Error(
      'No API key available. Please ensure you have configured your Bungie API key in DIM developer settings.',
    );
  }

  // Create HTTP client using shared implementation
  const httpClient = createBungieHttpClient(bungieApiKey, accessToken);

  return { httpClient, accessToken, membershipId };
}
