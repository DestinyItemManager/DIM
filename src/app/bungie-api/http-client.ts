import { delay } from 'app/utils/util';
import { PlatformErrorCodes, ServerResponse } from 'bungie-api-ts/destiny2';
import { HttpClient, HttpClientConfig } from 'bungie-api-ts/http';

/**
 * an error indicating a non-200 response code
 */
export class HttpStatusError extends Error {
  status: number;
  constructor(response: Response) {
    super(response.statusText);
    this.status = response.status;
  }
}

/**
 * an error indicating the Bungie API sent back a parseable response,
 * and that response indicated the request was not successful
 */
export class BungieError extends Error {
  code?: PlatformErrorCodes;
  status?: string;
  endpoint: string;
  constructor(response: Partial<ServerResponse<any>>, request: Request) {
    super(response.Message);
    this.name = 'BungieError';
    this.code = response.ErrorCode;
    this.status = response.ErrorStatus;
    this.endpoint = request.url;
  }
}

/**
 * this is a non-affecting pass-through for successful http requests,
 * but throws JS errors for a non-200 response
 */
function throwHttpError(response: Response) {
  if (response.status < 200 || response.status >= 400) {
    throw new HttpStatusError(response);
  }
  return response;
}

/**
 * sometimes what you have looks like a Response but it's actually an Error
 *
 * this is a non-affecting pass-through for successful API interactions,
 * but throws JS errors for "successful" fetches with Bungie error information
 */
function throwBungieError<T>(serverResponse: ServerResponse<T>, request: Request) {
  // There's an alternate error response that can be returned during maintenance
  const eMessage =
    serverResponse && (serverResponse as any).error && (serverResponse as any).error_description;
  if (eMessage) {
    throw new BungieError(
      {
        Message: eMessage,
        ErrorCode: PlatformErrorCodes.DestinyUnexpectedError,
        ErrorStatus: eMessage,
      },
      request
    );
  }

  if (serverResponse.ErrorCode !== PlatformErrorCodes.Success) {
    throw new BungieError(serverResponse, request);
  }

  return serverResponse;
}

//
// FETCH UTILS
//

/**
 * returns a fetch-like that will abort the request after some time
 *
 * @param fetchFunction use this function to make the request
 * @param timeout abort request after this many milliseconds
 */
export function createFetchWithTimeout(fetchFunction: typeof fetch, timeout: number): typeof fetch {
  return async (...[input, init]: Parameters<typeof fetch>) => {
    const controller = typeof AbortController === 'function' ? new AbortController() : null;
    const signal = controller?.signal;
    let timer: NodeJS.Timeout | undefined = undefined;

    if (controller) {
      timer = setTimeout(() => controller.abort(), timeout);
      if (typeof input === 'string') {
        input = new Request(input);
      }
      init = { ...init, signal };
    }
    try {
      return await fetchFunction(input, init);
    } finally {
      if (timer !== undefined) {
        clearTimeout(timer);
      }
    }
  };
}

/**
 * returns a fetch-like that will run a function if the request is taking a long time,
 * e.g. generate a "still waiting!" notification
 *
 * @param fetchFunction use this function to make the request
 * @param timeout run onTimeout after this many milliseconds
 * @param onTimeout the request's startTime and timeout will be passed to this
 */
export function createFetchWithNonStoppingTimeout(
  fetchFunction: typeof fetch,
  timeout: number,
  onTimeout: (startTime: number, timeout: number) => void
): typeof fetch {
  return async (...[input, init]: Parameters<typeof fetch>) => {
    const startTime = Date.now();
    const timer = setTimeout(() => onTimeout(startTime, timeout), timeout);

    try {
      return await fetchFunction(input, init);
    } finally {
      if (timer !== undefined) {
        clearTimeout(timer);
      }
    }
  };
}

//
// HTTPCLIENT UTILS
//

export function createHttpClient(
  fetchFunction: typeof fetch,
  apiKey: string,
  withCredentials: boolean
): HttpClient {
  return async (config: HttpClientConfig) => {
    let url = config.url;
    if (config.params) {
      // strip out undefined params keys. bungie-api-ts creates them for optional endpoint parameters
      for (const key in config.params) {
        typeof config.params[key] === 'undefined' && delete config.params[key];
      }
      url = `${url}?${new URLSearchParams(config.params).toString()}`;
    }

    const fetchOptions = new Request(url, {
      method: config.method,
      body: config.body ? JSON.stringify(config.body) : undefined,
      headers: { 'X-API-Key': apiKey, ...(config.body && { 'Content-Type': 'application/json' }) },
      credentials: withCredentials ? 'include' : 'omit',
    });
    const response = await fetchFunction(fetchOptions);
    const data: ServerResponse<any> = await response.json();
    // try throwing bungie errors, which have more information, first
    throwBungieError(data, fetchOptions);
    // then throw errors on generic http error codes
    throwHttpError(response);
    return data;
  };
}

let timesThrottled = 0;
/**
 * accepts an HttpClient and returns it with added throttling. throttles by increasing amounts
 * as it encounters Bungie API responses that indicate we should back off the requests, and
 * passes any thrown errors upstream
 *
 * @param httpClient use this client to make the API request
 * @param onThrottle run this when throttling happens. information about the throttling is passed in
 */
export function responsivelyThrottleHttpClient(
  httpClient: HttpClient,
  onThrottle: (timesThrottled: number, waitTime: number, url: string) => void
): HttpClient {
  return async (config: HttpClientConfig) => {
    if (timesThrottled > 0) {
      // Double the wait time, starting with 1 second, until we reach 5 minutes.
      const waitTime = Math.min(5 * 60 * 1000, Math.pow(2, timesThrottled) * 500);
      onThrottle(timesThrottled, waitTime, config.url);
      await delay(waitTime);
    }

    try {
      const result = await httpClient(config);
      // Quickly heal from being throttled
      timesThrottled = Math.floor(timesThrottled / 2);
      return result;
    } catch (e) {
      switch ((e as BungieError).code) {
        case PlatformErrorCodes.ThrottleLimitExceededMinutes:
        case PlatformErrorCodes.ThrottleLimitExceededMomentarily:
        case PlatformErrorCodes.ThrottleLimitExceededSeconds:
        case PlatformErrorCodes.DestinyThrottledByGameServer:
        case PlatformErrorCodes.PerApplicationThrottleExceeded:
        case PlatformErrorCodes.PerApplicationAnonymousThrottleExceeded:
        case PlatformErrorCodes.PerApplicationAuthenticatedThrottleExceeded:
        case PlatformErrorCodes.PerUserThrottleExceeded:
        case PlatformErrorCodes.SystemDisabled:
          timesThrottled++;
          break;
      }
      throw e;
    }
  };
}
