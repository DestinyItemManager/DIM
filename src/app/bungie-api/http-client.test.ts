import {
  DestinyComponentType,
  getItem,
  getLinkedProfiles,
  getProfile,
  getVendor,
  pullFromPostmaster,
} from 'bungie-api-ts/destiny2';
import { createHttpClient } from './http-client';

const errors = {
  DestinyNoRoomInDestination: {
    Response: 0,
    ErrorCode: 1642,
    ThrottleSeconds: 0,
    ErrorStatus: 'DestinyNoRoomInDestination',
    Message: 'There are no item slots available to transfer this item.',
    MessageData: {},
  },
  SystemDisabled: {
    Response: 0,
    ErrorCode: 5,
    ThrottleSeconds: 0,
    ErrorStatus: 'SystemDisabled',
    Message: 'This system is temporarily disabled for maintenance.',
    MessageData: {},
  },
};

type TroubleshootingResponse = { req: Request; ErrorCode: number };

const makePretendFetch = (response?: any) => (req: any) => ({
  json: () => ({ req: req as Request, ErrorCode: 1, ...response }),
});
const pretendHttpClient = (response?: any) =>
  createHttpClient(makePretendFetch(response) as any as typeof fetch, '123');

const cases: [(...params: any) => any, object | undefined][] = [
  [
    getItem,
    {
      destinyMembershipId: '123456',
      membershipType: 3,
      itemInstanceId: '0987654321',
      components: [7],
    },
  ],
  [
    getProfile,
    {
      destinyMembershipId: '123456',
      membershipType: 3,
      components: [1, 2, 3],
    },
  ],
  [
    getLinkedProfiles,
    {
      membershipId: '123456',
      membershipType: 3,
      getAllMemberships: true,
    },
  ],

  [
    getVendor,
    {
      characterId: '1234658790',
      destinyMembershipId: '123456',
      membershipType: 3,
      components: [
        DestinyComponentType.Vendors,
        DestinyComponentType.VendorSales,
        DestinyComponentType.ItemInstances,
        DestinyComponentType.ItemObjectives,
        DestinyComponentType.ItemStats,
        DestinyComponentType.ItemSockets,
        DestinyComponentType.ItemTalentGrids,
        DestinyComponentType.ItemCommonData,
        DestinyComponentType.CurrencyLookups,
        DestinyComponentType.ItemPlugStates,
        DestinyComponentType.ItemReusablePlugs,
        // TODO: We should try to defer this until the popup is open!
        DestinyComponentType.ItemPlugObjectives,
      ],
      vendorHash: 45674576,
    },
  ],
  [
    pullFromPostmaster,
    {
      characterId: '1234658790',
      membershipType: 3,
      itemId: '0987654321',
      itemReferenceHash: 45674576,
      stackSize: 7,
      transferToVault: true,
    },
  ],
];

test.each(cases)('check Request builder for %p', async (apiFunc, apiFuncParams) => {
  const request: TroubleshootingResponse = await apiFunc(pretendHttpClient(), apiFuncParams);
  const { headers, method, url, body } = request.req;
  expect({ headers, method, url, body }).toMatchSnapshot();
});

test('should throw an error if there is no room in the destination', async () => {
  expect.assertions(1);
  await expect(async () => {
    (await pullFromPostmaster(pretendHttpClient(errors.DestinyNoRoomInDestination), {
      characterId: '1234658790',
      membershipType: 3,
      itemId: '0987654321',
      itemReferenceHash: 45674576,
      stackSize: 7,
    })) as any;
  }).rejects.toMatchInlineSnapshot(
    `[BungieError: There are no item slots available to transfer this item.]`,
  );
});

test('should throw an error if API is down for maintenance', async () => {
  expect.assertions(1);
  await expect(async () => {
    (await pullFromPostmaster(pretendHttpClient(errors.SystemDisabled), {
      characterId: '1234658790',
      membershipType: 3,
      itemId: '0987654321',
      itemReferenceHash: 45674576,
      stackSize: 7,
    })) as any;
  }).rejects.toMatchInlineSnapshot(
    `[BungieError: This system is temporarily disabled for maintenance.]`,
  );
});
