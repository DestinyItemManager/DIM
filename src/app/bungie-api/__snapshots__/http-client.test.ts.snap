// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`check Request builder for [Function getItem] 1`] = `
{
  "body": null,
  "headers": Headers {
    Symbol(map): {
      "X-API-Key": [
        "123",
      ],
    },
  },
  "method": "GET",
  "url": "https://www.bungie.net/Platform/Destiny2/3/Profile/123456/Item/0987654321/?components=7",
}
`;

exports[`check Request builder for [Function getLinkedProfiles] 1`] = `
{
  "body": null,
  "headers": Headers {
    Symbol(map): {
      "X-API-Key": [
        "123",
      ],
    },
  },
  "method": "GET",
  "url": "https://www.bungie.net/Platform/Destiny2/3/Profile/123456/LinkedProfiles/?getAllMemberships=true",
}
`;

exports[`check Request builder for [Function getProfile] 1`] = `
{
  "body": null,
  "headers": Headers {
    Symbol(map): {
      "X-API-Key": [
        "123",
      ],
    },
  },
  "method": "GET",
  "url": "https://www.bungie.net/Platform/Destiny2/3/Profile/123456/?components=1%2C2%2C3",
}
`;

exports[`check Request builder for [Function getVendor] 1`] = `
{
  "body": null,
  "headers": Headers {
    Symbol(map): {
      "X-API-Key": [
        "123",
      ],
    },
  },
  "method": "GET",
  "url": "https://www.bungie.net/Platform/Destiny2/3/Profile/123456/Character/1234658790/Vendors/45674576/?components=400%2C402%2C300%2C301%2C304%2C305%2C306%2C307%2C600%2C308%2C310%2C309",
}
`;

exports[`check Request builder for [Function pullFromPostmaster] 1`] = `
{
  "body": {
    "data": [
      123,
      34,
      99,
      104,
      97,
      114,
      97,
      99,
      116,
      101,
      114,
      73,
      100,
      34,
      58,
      34,
      49,
      50,
      51,
      52,
      54,
      53,
      56,
      55,
      57,
      48,
      34,
      44,
      34,
      109,
      101,
      109,
      98,
      101,
      114,
      115,
      104,
      105,
      112,
      84,
      121,
      112,
      101,
      34,
      58,
      51,
      44,
      34,
      105,
      116,
      101,
      109,
      73,
      100,
      34,
      58,
      34,
      48,
      57,
      56,
      55,
      54,
      53,
      52,
      51,
      50,
      49,
      34,
      44,
      34,
      105,
      116,
      101,
      109,
      82,
      101,
      102,
      101,
      114,
      101,
      110,
      99,
      101,
      72,
      97,
      115,
      104,
      34,
      58,
      52,
      53,
      54,
      55,
      52,
      53,
      55,
      54,
      44,
      34,
      115,
      116,
      97,
      99,
      107,
      83,
      105,
      122,
      101,
      34,
      58,
      55,
      44,
      34,
      116,
      114,
      97,
      110,
      115,
      102,
      101,
      114,
      84,
      111,
      86,
      97,
      117,
      108,
      116,
      34,
      58,
      116,
      114,
      117,
      101,
      125,
    ],
    "type": "Buffer",
  },
  "headers": Headers {
    Symbol(map): {
      "Content-Type": [
        "application/json",
      ],
      "X-API-Key": [
        "123",
      ],
    },
  },
  "method": "POST",
  "url": "https://www.bungie.net/Platform/Destiny2/Actions/Items/PullFromPostmaster/",
}
`;
